tapdance/dns-registrar.go

```diff
@@ -0,0 +1,259 @@
+package tapdance
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"net"
+	"time"
+
+	"github.com/ccding/go-stun/stun"
+	"github.com/golang/protobuf/proto"
+	"github.com/refraction-networking/gotapdance/pkg/dns-registrar/requester"
+	pb "github.com/refraction-networking/gotapdance/protobuf"
+)
+
+type DNSRegistrar struct {
+	req             *requester.Requester
+	maxTries        int
+	connectionDelay time.Duration
+	bidirectional   bool
+	ip              []byte
+}
+
+// Create a DNSRegistrar from DnsRegConf.
+// Uses the pubkey in conf as default. If it is not supplied (nil), uses fallbackKey instead.
+func NewDNSRegistrarFromConf(conf *pb.DnsRegConf, bidirectional bool, delay time.Duration, maxTries int, fallbackKey []byte) (*DNSRegistrar, error) {
+	pubkey := conf.Pubkey
+	if pubkey == nil {
+		pubkey = fallbackKey
+	}
+	udpAddr, dotAddr, dohUrl := "", "", ""
+	switch *conf.DnsRegMethod {
+	case pb.DnsRegMethod_UDP:
+		udpAddr = *conf.UdpAddr
+	case pb.DnsRegMethod_DOT:
+		dotAddr = *conf.DotAddr
+	case pb.DnsRegMethod_DOH:
+		dohUrl = *conf.DohUrl
+	default:
+		return nil, errors.New("unkown reg method in conf")
+	}
+	return NewDNSRegistrar(udpAddr, dotAddr, dohUrl, *conf.Domain, pubkey, *conf.UtlsDistribution, maxTries, bidirectional, delay, *conf.StunServer)
+}
+
+// Create a DNSRegistrar. Exactly one of udpAddr, dotAddr, and dohUrl should be provided.
+func NewDNSRegistrar(udpAddr string, dotAddr string, dohUrl string, domain string, pubkey []byte, utlsDistribution string, maxTries int, bidirectional bool, delay time.Duration, stun_server string) (*DNSRegistrar, error) {
+	r := &DNSRegistrar{}
+	r.maxTries = maxTries
+	r.bidirectional = bidirectional
+	r.connectionDelay = delay
+	var err error
+	if utlsDistribution == "" {
+		utlsDistribution = "3*Firefox_65,1*Firefox_63,1*iOS_12_1"
+	}
+	if domain == "" {
+		return nil, errors.New("domain must be specified")
+	}
+	if pubkey == nil {
+		return nil, errors.New("server public key must be provided")
+	}
```

Comment:
Is there a reason not to make this 

```
	if udpAddr != "" {
	...
	} else if dohUrl != "" {
	...
	} else if dotUrl != "" {
	...
	}

	if r.req == nil {
		return nil, errors.New("a valid valid value for one of udpAddr, dohUrl, or dotAddr must be provided")
	}

Reply:
It is so that if more than one of the transport options are given, it will error out saying only one should be given.

Reply:
Ah that make a lot more sense. Changed it in the latest commit.